pymc_experimental/utils/linear_cg.py

```diff
@@ -0,0 +1,279 @@
+
+
+class settings:
+    eval_cg_tolerance = 0.01
+    cg_tolerance = 1
+
+    def terminate_cg_by_size_on():
+        return False
+
+    def _use_eval_tolerance_on():
+        return False
+
+    terminate_cg_by_size = Setting(on=terminate_cg_by_size_on)
+    _use_eval_tolerance = Setting(on=_use_eval_tolerance_on)
+
+
+def masked_fill(vector, mask, fill_value):
+    masked_vector = np.ma.array(vector, mask=mask)
+    vector = masked_vector.filled(fill_value=fill_value)
+    return vector
+
+
+def linear_cg_updates(
+    result, alpha, residual_inner_prod, eps, beta, residual, precond_residual, curr_conjugate_vec
+):
+
+    # Everything inside _jit_linear_cg_updates
+    result = result + alpha * curr_conjugate_vec
+    beta = np.copy(residual_inner_prod)
+
+    residual_inner_prod = residual.T @ precond_residual
+
+    # safe division
+    is_zero = beta < eps
+    beta = masked_fill(beta, mask=is_zero, fill_value=1)
+
+    beta = residual_inner_prod / beta
+    beta = masked_fill(beta, mask=is_zero, fill_value=0)
+    curr_conjugate_vec = beta * curr_conjugate_vec + precond_residual
+    return (
+        result,
+        alpha,
+        residual_inner_prod,
+        eps,
+        beta,
+        residual,
+        precond_residual,
+        curr_conjugate_vec,
+    )
+
+
+def linear_cg(
+    mat: np.matrix,
+    rhs,
+    n_tridiag=0,
+    tolerance=None,
+    eps=1e-10,
+    stop_updating_after=1e-10,
+    max_iter=1000,
+    max_tridiag_iter=20,
+    initial_guess=None,
+    preconditioner=None,
+):
+
+    if initial_guess is None:
+        initial_guess = np.zeros_like(rhs)
+
+    if preconditioner is None:
+        preconditioner = lambda x: x
+        precond = False
+    else:
+        precond = True
+
+    if tolerance is None:
+        if settings._use_eval_tolerance.on():
+            tolerance = settings.eval_cg_tolerance
+        else:
+            tolerance = settings.cg_tolerance
+
+    # If we are running m CG iterations, we obviously can't get more than m Lanczos coefficients
+    if max_tridiag_iter > max_iter:
+        raise RuntimeError(
+            "Getting a tridiagonalization larger than the number of CG iterations run is not possible!"
+        )
+
+    is_vector = len(rhs.shape) == 1
+    if is_vector:
+        rhs = rhs[:, np.newaxis]
+
+    num_rows = rhs.size
+    n_iter = min(max_iter, num_rows) if settings.terminate_cg_by_size.on() else max_iter
+    n_tridiag_iter = min(max_tridiag_iter, num_rows)
+
+    # norm of rhs for convergence tests
+    rhs_norm = np.linalg.norm(rhs, 2)
+    # make almost-zero norms be 1 (so we don't get divide-by-zero errors)
+    rhs_is_zero = rhs_norm < eps
+    rhs_norm = masked_fill(rhs_norm, mask=rhs_is_zero, fill_value=1)
+
+    # lets normalize rhs
+    rhs = rhs / rhs_norm
+
+    # residuals
+    residual = rhs - mat @ initial_guess
+    batch_shape = residual.shape[:-2]
+
+    result = np.copy(initial_guess)
+
+    if np.allclose(residual, residual):
```

Comment:
Just a heads up, I noticed this check fail yesterday. Was passing when I ran it on Colab, looking into it.

Reply:
Fixed it should have been `if not np.allclose(...)`